api/migrations/20241105100553-add-cohort-groups.js

```diff
@@ -0,0 +1,67 @@
+import { CohortGroupModel, CohortModel } from "../../../../src/models";
+import { COHORT_TYPE } from "snu-lib";
+
+const pipeline = [
+  {
+    $addFields: {
+      year: { $year: "$dateStart" },
+    },
+  },
+  {
+    $group: {
+      _id: { type: "$type", year: "$year" },
+      documents: {
+        $push: { _id: "$_id", name: "$name" },
+      },
+    },
+  },
+  {
+    $project: {
+      _id: 0,
+      type: "$_id.type",
+      year: "$_id.year",
+      documents: 1,
+    },
+  },
+];
+
+const fromUser = { firstName: "Migration ajout des groupes de cohortes" };
+
+module.exports = {
+  async up() {
+    try {
+      const reserveGroup = new CohortGroupModel({ name: "Réserve" });
+      await reserveGroup.save({ fromUser });
+
+      const groups = await CohortModel.aggregate(pipeline);
+
+      for (const group of groups) {
```

Comment:
Je sais que c'est un fichier de migration mais pour ce cas avec les boucles for avec des instructions findBy* c'est un peu couteux en I/O et bloque le parallélisme:
- Utiliser les requêtes en masse:
```
const allCohortIds = groups.flatMap(group => group.documents.map(doc => doc._id)).filter(id => id);
const cohorts = await CohortModel.find({ _id: { $in: allCohortIds } });
const cohortMap = new Map(cohorts.map(cohort => [cohort._id.toString(), cohort]));
```
- utiliser le `insertMany`
- Paralléliser les opérations

Par exemple:
```javascript
try {
      // 1. Créer le groupe de réserve
      const reserveGroup = new CohortGroupModel({ name: "Réserve" });
      await reserveGroup.save({ fromUser });
      logger.info("Groupe de réserve créé avec succès.");

      // 2. Agréger les groupes existants
      const groups = await CohortModel.aggregate(pipeline);
      logger.info(`Agrégation terminée. Nombre de groupes trouvés: ${groups.length}`);

      if (groups.length === 0) {
        logger.info("Aucun groupe trouvé.");
        return;
      }

      // 3. Préparer les nouveaux groupes à insérer
      const newGroups = groups.map(group => {
        const { type, year } = group;
        const name = `${type === COHORT_TYPE.VOLONTAIRE ? "HTS" : "CLE"} ${year}`;
        return { type, year, name };
      });

      // 4. Insérer tous les nouveaux groupes en une seule opération
      const insertedGroups = await CohortGroupModel.insertMany(newGroups, { ordered: false });
      logger.info(`Insertion des nouveaux groupes terminée. Nombre de groupes insérés: ${insertedGroups.length}`);

      // 5. Créer une map des groupes insérés pour un accès rapide
      const groupMap = new Map(insertedGroups.map(group => [group.name, group._id]));
      logger.info("Map des groupes insérés créée.");

      // 6. Collecter tous les IDs de cohortes pour une récupération en masse + Filtrer les IDs valides
      const allCohortIds = groups
        .flatMap(group => group.documents.map(doc => doc._id))
        .filter(id => id); 

      // 7. Récupérer toutes les cohortes en une seule requête
      const cohorts = await CohortModel.find({ _id: { $in: allCohortIds } });
      logger.info(`Récupération des cohortes terminée. Nombre de cohortes trouvées: ${cohorts.length}`);

      // 8. Préparer les opérations de mise à jour
      const cohortUpdates = cohorts.map(cohort => {
        const groupName = `${cohort.type === COHORT_TYPE.VOLONTAIRE ? "HTS" : "CLE"} ${new Date(cohort.dateStart).getFullYear()}`;
        const targetGroupId = cohort.name === "à venir" ? reserveGroup._id : groupMap.get(groupName);

        if (!targetGroupId) {
          throw new Error(`Aucun groupe cible trouvé pour la cohorte: ${cohort._id}`);
        }

        // Définir le champ cohortGroupId
        cohort.cohortGroupId = targetGroupId;

        return cohort.save({ fromUser });
      });

      // 9. Exécuter toutes les mises à jour en parallèle
      await Promise.all(cohortUpdates);
      logger.info("Mises à jour des cohortes terminées avec succès.");

    } catch (error) {
      logger.error("Migration échouée:", error);
      throw error;
    }
    ```

Reply:
Question au sujet du insertMany sur les groupes : on perd l'info de l'auteur de la modif non ? (La propriété "fromUser" fournie à la fonction save(), qui sert à renseigner l'info user dans les patches).

Reply:
Autre question : pourquoi throw dans le catch ? C'est pour la lib mongo-migrate ? Pourquoi le try catch du coup ?

Reply:
> Autre question : pourquoi throw dans le catch ? C'est pour la lib mongo-migrate ? Pourquoi le try catch du coup ? A-t-on besoin de logger une erreur en plus de l'échec de la migration ?
Logger l'erreur pour les migrations c'est pour la remonter avec winston; le throw c'est plus par habitude de mon côté, effectivement, tu peux supprimer le `throw error`.

Reply:
> Question au sujet du insertMany sur les groupes : on perd l'info de l'auteur de la modif non ? (La propriété "fromUser" fournie à la fonction save(), qui sert à renseigner l'info user dans les patches).

Pour le insertMany sur les groupes tu peux rajouter le fromUser au niveau du map, mais je vois que tu l'as fait sur ta dernière version, tout est good pour moi top 👌